internal/state/graph/backend_refs.go

```diff
@@ -11,19 +11,33 @@ import (
 	"github.com/nginxinc/nginx-kubernetes-gateway/internal/state/conditions"
 )
 
-func addBackendGroupsToRoutes(
+// BackendRef is an internal representation of a backendRef in an HTTPRoute.
+type BackendRef struct {
+	// Svc is the service referenced by the backendRef.
+	Svc *v1.Service
+	// Name is the name of the backendRef.
+	Name string
```

Comment:
> Perhaps we can just create a unique upstream names in the dataplane package?

We can, it will just move the responsibility of deduplicating and naming to the dataplane package.

Reply:
I still think the name field is a bit problematic. For example, consider two backend refs to the same service and the same port but with different weight. For those, NKG will generate two different BackendRefs with the same name. So the field is not really a name - it is a unique reference to the service port.

Also, the name only gets generated is the service is resolved.

So perhaps we can remove this field completely and let the data plane package construct a proper upstream name based on the BackendRef? (It can expect the Svc field and make a decision)

Reply:
>  So the field is not really a name - it is a unique reference to the service port.

It's really the upstream name.


If we remove the name field, then we have to construct the upstream name twice: once when [building upstreams ](https://github.com/nginxinc/nginx-kubernetes-gateway/blob/479dc09d8b8a12ed6f819f4f3c7a03d3074c9325/internal/state/dataplane/configuration.go#L386) and once when [creating the backend group](https://github.com/nginxinc/nginx-kubernetes-gateway/blob/479dc09d8b8a12ed6f819f4f3c7a03d3074c9325/internal/state/dataplane/configuration.go#L205). 

I think that's why I originally stored the name in BackendRef, so it only had to be generated once and there was no risk that the Upstream.Name and Backend.Name would differ.

Reply:
@kate-osborn

> It's really the upstream name.

maybe some renaming can be helpful here. For example, `ServicePortReference`,  because we only set the `name` field to something if the backend ref references a service and when the service exists.

would it also make sense to generate that value rather than store it generated?  Because we store a Service reference in the BackendRef, we can just use that data in a method to generate a value . This way, we will have a single source of truth. What do you think?

Reply:
`ServicePortReference` sounds good.

> would it also make sense to generate that value rather than store it generated?

Similar to how `BackendGroup.Name()` works? Something like `BackendRef.ServicePortReference()`?